src/Microsoft.AspNet.Mvc.DataAnnotations/DataAnnotationsModelValidator.cs

```diff
@@ -26,6 +27,115 @@ public DataAnnotationsModelValidator(ValidationAttribute attribute, IStringLocal
+                    {
+                        return new RegularExpressionAttributeAdapter((RegularExpressionAttribute)Attribute, _stringLocalizer)
+                            .GetErrorMessage(modelMetadata);
+                    }
+                },
+                {
+                    typeof(CompareAttribute), (modelMetadata) =>
+                    {
+                        return new CompareAttributeAdapter((CompareAttribute)Attribute, _stringLocalizer)
+                            .GetErrorMessage(modelMetadata, metadataProvider);
+
+                    }
+                },
+                {
+                    typeof(StringLengthAttribute), (modelMetadata) =>
+                    {
+                        return new StringLengthAttributeAdapter((StringLengthAttribute)Attribute, _stringLocalizer)
+                            .GetErrorMessage(modelMetadata);
+                    }
+                },
+                {
+                    typeof(CreditCardAttribute), (modelMetadata) =>
+                    {
+                        return new DataTypeAttributeAdapter((DataTypeAttribute)Attribute, "creditcard", _stringLocalizer)
+                            .GetErrorMessage(modelMetadata);
+                    }
+                },
+                {
+                    typeof(EmailAddressAttribute), (modelMetadata) =>
+                    {
+                        return new DataTypeAttributeAdapter((DataTypeAttribute)Attribute, "email", _stringLocalizer)
+                            .GetErrorMessage(modelMetadata);
+                    }
+                },
+                {
+                    typeof(PhoneAttribute), (modelMetadata) =>
+                    {
+                        return new DataTypeAttributeAdapter((DataTypeAttribute)Attribute, "phone", _stringLocalizer)
+                            .GetErrorMessage(modelMetadata);
+                    }
+                },
+                {
+                    typeof(UrlAttribute), (modelMetadata) =>
+                    {
+                        return new DataTypeAttributeAdapter((DataTypeAttribute)Attribute, "url", _stringLocalizer)
+                            .GetErrorMessage(modelMetadata);
+                    }
+                }
+            };
+        }
+
+        private string GetErrorMessage(ModelMetadata metadata, IModelMetadataProvider metadataProvider)
+        {
+            if (Attribute == null)
+            {
+                throw new ArgumentNullException(nameof(Attribute));
+            }
+
+
+            var attrType = Attribute.GetType();
+
+            var dict = BuildErrorMessageDict(metadataProvider);
+
+            if (dict.ContainsKey(attrType))
+            {
+                var func = dict[attrType];
+                return func(metadata);
+            }
+            else
+            {
+                throw new NotImplementedException($"Error message localizer not defined for {attrType.Name}");
```

Comment:
Need to find a way to silently ignore the missing adapter, which is what client side does today.